# Week 3 - Data Wrangling

In this notebook we will practice data cleaning and wrangling using a *messy* version of the Titanic dataset.

Topics covered:
- Inspecting data
- Preprocessing
    - Categorical x Numerical Data
    - Fixing column types
    - Standardizing Categorical Values
    - Missing values (Identifying and Imputation)
- Feature Engineering
- Exporting and Loading Cleaned Data

### SETUP

In [140]:
# if you any of those
#!pip install pandas numpy seaborn

In [141]:
import pandas as pd
import numpy as np
import seaborn as sns

##  Dataset

In this section, we will work with real-world datasets to apply the concepts we have learned so far.

Let's load a sample dataset using Pandas.

We are going to use the popular `Titanic` dataset, which contains information about passengers on the Titanic, including whether they survived the disaster.

**Data dictionary**

| Column | Description |
|---|---|
| PassengerId | Unique ID : T (Titanic) + ticket class + year + port + incremental ID per class.
| Survived | 0 = No, 1 = Yes |
| Pclass | Ticket class (1, 2, 3) |
| Name | Passenger name |
| Sex | Gender |
| Age | Age in years |
| SibSp | Siblings/spouses aboard |
| Parch | Parents/children aboard |
| Fare | Ticket fare |
| Deck | Deck number |
| Embarked | Port of embarkation (C/Q/S)|


### Load the Titanic dataset:

In [142]:
df = pd.read_csv('messy_titanic.csv')
df

;survived;pclass;sex;age;sibsp;parch;fare;adult_male;deck;embark_town;passengerId;name
0    0;0;3;Male;22.0;1;0;7.25;True;;Southampton;T3-...                                    
1    1;1;1;female;38.0;1;0;71.2833;False;C;Cherbour...                                    
2    2;1;3;female;26.0;0;0;7.925;False;;Southampton...                                    
3    3;1;1;female;35.0;1;0;53.1;False;C;Southampton...                                    
4    4;0;3;Male;35.0;0;0;8.05;True;;Southampton;T3-...                                    
..                                                 ...                                    
886  886;0;2;Male;27.0;0;0;13.0;True;;Southampton;T...                                    
887  887;1;1;female;19.0;0;0;30.0;False;B;Southampt...                                    
888  888;0;3;female;nan;1;2;23.45;False;;Southampto...                                    
889  889;1;1;Male;26.0;0;0;;True;C;Cherbourg;T1-191...                                    
890  890;0;3;Male;32.0;0;0;7.75;True;;Queenstown;T3...                                    

[891 rows x 1 columns]

Before anything, we must load the dataset correctly

Manually inspect the dataset file or run the command `!head file_name.csv` or `!cat file_name.csv` - **head** might not work in windows OS 😊

In [143]:
# the delimiter being used is ;
!cat messy_titanic.csv

;survived;pclass;sex;age;sibsp;parch;fare;adult_male;deck;embark_town;passengerId;name
0;0;3;Male;22.0;1;0;7.25;True;;Southampton;T3-1912-S-P0001;Dustin Flores
1;1;1;female;38.0;1;0;71.2833;False;C;Cherbourg;T1-1912-C-P0001;Mary Tyler
2;1;3;female;26.0;0;0;7.925;False;;Southampton;T3-1912-S-P0002;Anna Mendoza
3;1;1;female;35.0;1;0;53.1;False;C;Southampton;T1-1912-S-P0001;Debra Pearson
4;0;3;Male;35.0;0;0;8.05;True;;Southampton;T3-1912-S-P0003;Anthony Taylor
5;0;3;Male;nan;0;0;8.4583;True;;Queenstown;T3-1912-Q-P0001;Eric Mayo
6;0;1;Male;54.0;0;0;51.8625;True;E;Southampton;T1-1912-S-P0002;Brian James
7;0;3;Male;2.0;3;1;21.075 USD;False;;Southampton;T3-1912-S-P0004;Tyler Smith
8;1;3;female;27.0;0;2;11.1333 USD;False;;Southampton;T3-1912-S-P0005;Robin Dodson
9;1;2;female;14.0;1;0;30.0708;False;;Cherbourg;T2-1912-C-P0001;Sandra Ward
10;1;3;female;4.0;1;1;16.7 USD;False;G;Southampton;T3-1912-S-P0006;Kimberly Stevenson DDS
11;1;1;female;58.0;0;0;26.55 USD;False;C;Southampton;T1-1912-S-P0003;M

Now try again using `sep`

In [144]:
df = pd.read_csv('messy_titanic.csv', sep=';')
df

Unnamed: 0  survived  pclass     sex   age  sibsp  parch     fare  \
0             0         0       3    Male  22.0      1      0     7.25   
1             1         1       1  female  38.0      1      0  71.2833   
2             2         1       3  female  26.0      0      0    7.925   
3             3         1       1  female  35.0      1      0     53.1   
4             4         0       3    Male  35.0      0      0     8.05   
..          ...       ...     ...     ...   ...    ...    ...      ...   
886         886         0       2    Male  27.0      0      0     13.0   
887         887         1       1  female  19.0      0      0     30.0   
888         888         0       3  female   NaN      1      2    23.45   
889         889         1       1    Male  26.0      0      0      NaN   
890         890         0       3    Male  32.0      0      0     7.75   

     adult_male deck  embark_town      passengerId                 name  
0          True  NaN  Southampton  T3-1912-S-P0001        Dustin Flores  
1         False    C    Cherbourg  T1-1912-C-P0001           Mary Tyler  
2         False  NaN  Southampton  T3-1912-S-P0002         Anna Mendoza  
3         False    C  Southampton  T1-1912-S-P0001        Debra Pearson  
4          True  NaN  Southampton  T3-1912-S-P0003       Anthony Taylor  
..          ...  ...          ...              ...                  ...  
886        True  NaN  Southampton  T2-1912-S-P0164      Michael Johnson  
887       False    B  Southampton  T1-1912-S-P0127            Lisa Buck  
888       False  NaN  Southampton  T3-1912-S-P0353  Alexandria Lawrence  
889        True    C    Cherbourg  T1-1912-C-P0085     Christopher Reed  
890        True  NaN   Queenstown  T3-1912-Q-P0072         James Horton  

[891 rows x 13 columns]

**It seems that someone saved the index by mistake... Lets get rid of it!**

### Q.1.1 Drop the unnamed column.
- Use `df.drop(columns=['col_name'], inplace=True)`

In [145]:
df.drop(columns=['Unnamed: 0'], inplace=True)
df.head()

survived  pclass     sex   age  sibsp  parch     fare  adult_male deck  \
0         0       3    Male  22.0      1      0     7.25        True  NaN   
1         1       1  female  38.0      1      0  71.2833       False    C   
2         1       3  female  26.0      0      0    7.925       False  NaN   
3         1       1  female  35.0      1      0     53.1       False    C   
4         0       3    Male  35.0      0      0     8.05        True  NaN   

   embark_town      passengerId            name  
0  Southampton  T3-1912-S-P0001   Dustin Flores  
1    Cherbourg  T1-1912-C-P0001      Mary Tyler  
2  Southampton  T3-1912-S-P0002    Anna Mendoza  
3  Southampton  T1-1912-S-P0001   Debra Pearson  
4  Southampton  T3-1912-S-P0003  Anthony Taylor

### Q.1.2 How many rows and columns there is?

In [146]:
# 891 rows and 12 columns
df.shape


(891, 12)

### Q.1.3 Check of each column data type

- Is every column data type the most appropriate for it?
- Are there any column with null/missing values?

In [147]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         846 non-null    object 
 7   adult_male   891 non-null    bool   
 8   deck         203 non-null    object 
 9   embark_town  889 non-null    object 
 10  passengerId  891 non-null    object 
 11  name         891 non-null    object 
dtypes: bool(1), float64(1), int64(4), object(6)
memory usage: 77.6+ KB


### Q.1.4 Verify descritive statistics
- All numeric columns are being displayed?
- Those descritive statistcs are meaningfull for every column?

In [148]:
df.describe()

survived      pclass         age       sibsp       parch
count  891.000000  891.000000  714.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594
std      0.486592    0.836071   14.526497    1.102743    0.806057
min      0.000000    1.000000    0.420000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000
50%      0.000000    3.000000   28.000000    0.000000    0.000000
75%      1.000000    3.000000   38.000000    1.000000    0.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000

## 2. Preprocessing

Now that we have at least some information about the dataset, it is more than obvious that there are many problems.

Let's preprocess this dataset to make it more viable and clean for future analysis.

### Q.2.1 First of all, the documentation provided a column name that does not exist.
- Identify and `rename` it:

**Example:**

```python
df.rename(columns={'old_name': 'new_name'}, inplace=True)
```

In [ ]:
# lets rename deck to deck
df.rename(columns={'deck': 'deck'}, inplace=True)
df.head()

survived  pclass     sex   age  sibsp  parch     fare  adult_male cabin  \
0         0       3    Male  22.0      1      0     7.25        True   NaN   
1         1       1  female  38.0      1      0  71.2833       False     C   
2         1       3  female  26.0      0      0    7.925       False   NaN   
3         1       1  female  35.0      1      0     53.1       False     C   
4         0       3    Male  35.0      0      0     8.05        True   NaN   

   embark_town      passengerId            name  
0  Southampton  T3-1912-S-P0001   Dustin Flores  
1    Cherbourg  T1-1912-C-P0001      Mary Tyler  
2  Southampton  T3-1912-S-P0002    Anna Mendoza  
3  Southampton  T1-1912-S-P0001   Debra Pearson  
4  Southampton  T3-1912-S-P0003  Anthony Taylor

<details>
<summary style="background-color: #fff3cd; padding: 10px; border-radius: 5px; cursor: pointer;">
    <h2 style="display: inline;">
        Understanding Numerical vs. Categorical Data
        <span style="color: red; font-size: 0.8em;">(click to expand/collapse)</span>
    </h2>
</summary>

It is important to understand **what type of data each column represents**.  
Choosing the **correct type** improves **memory usage**, **speeds up operations**, and makes **analysis more reliable**.

---

### **1. Numerical Data**
>
>These are values that represent quantities and support mathematical operations.
>
>**Examples:**
>- Age (`22`, `45.0`)
>- Fare (`71.283`, `15.50`)
>- Number of siblings (`3`, `0`)
>- Any measured or counted quantity
>
>**Two main subtypes:**
>- **int** → whole numbers (0, 1, 2…)
>- **float** → decimal numbers (0.0, 3.14…)
>
>**Numerical columns allow:**
>- statistics (mean, median, std)
>- plotting histograms
>- mathematical operations
>- machine learning models
---
### **2. Categorical Data**
>
>These represent **labels**, **groups**, or **categories**, not quantities.
>
>**Examples:**
>- Sex (`"male"`, `"female"`)
>- Embarkation port (`"C"`, `"Q"`, `"S"`)
>- Passenger class (`"First"`, `"Second"`, `"Third"`)
>- Deck (`"A"`, `"B"`, `"Unknown"`)
>
>**Types commonly used:**
>- **string (object)** → free text labels  
>- **category** → optimized categorical labels (recommended)
>
>**Categorical columns allow:**
>- grouping (`df.groupby("sex")`)
>- value counts
>- filtering
>- faster processing when stored as `category`
---
### **Why this distinction matters**
>
>If a column is stored with the wrong dtype:
>
>- A numeric column stored as string **cannot be computed**:  
>  `"45" + "10"` → `"4510"` (string concatenation!)
>- A categorical column stored as numbers **breaks meaning**
>- Machine learning pipelines **fail or behave poorly**
>- Missing values become difficult to detect and fix
>- Memory usage increases (especially with text)
>
>**So the first step in Data Wrangling is:**
>
>👉 **Identify** whether each column is numerical or categorical  
>👉 **Fix** the dtype accordingly
</details>

## Fixing Column Types

Datasets often contain columns stored with the wrong data type.  
This usually happens when values are mixed (e.g., numbers + text), when the file was exported incorrectly, or when data is collected from multiple sources.

Working with incorrect types can break calculations, comparisons, plots, or machine-learning pipelines.  
So the first step is to **inspect** and **fix** them.

**Common strategies for fixing the column types are:**

___
>**Direct convertion:**
>
>If you know that the column will be able to be directly converted
>```python
>df['column_name'] = df['column_name'].astype(dtype)
>```
___
**1. Convert a column to numeric**
>
>Use when the column **should be a number**, but contains strings, symbols, or inconsistent formatting.
>
>```python
>df["column_name"] = pd.to_numeric(df["column_name"], errors="coerce")
>```
>`errors="coerce"` converts invalid values into **NaN**, which can later be imputed.
___

**2. Clean and convert a column that contains units or extra text**
>
>Useful for columns like "100 USD", "45 years", "200 km".
>
>```python
>df["column_name"] = (
>    df["column_name"]
>        .astype(str)
>        .str.replace(" USD", "", regex=False)
>)
>
># now casting to numeric dtype
>df["column_name"] = pd.to_numeric(df["column_name"], errors="coerce")
>```
___
**3. Convert a column to string (categorical text)**
>
>Use when the column should represent labels, not numbers.
>
>```python
>df["column_name"] = df["column_name"].astype(str)
>```
___
**4. Convert a column to category (optional optimization)**
>
>Useful for columns with repeated labels (e.g., "male", "female", "unknown").
>
>```python
>df["column_name"] = df["column_name"].astype("category")
>```
___
**5. Convert a column to datetime**
>
>For dates like "2020-01-05" or "05/01/2020".
>```python
>df["column_name"] = pd.to_datetime(df["column_name"], errors="coerce")
>```
___
**After fixing all types, it is good practice to check the result:**
>
>- `.info()`
>- `.dtypes`

### Q 2.2 Inspect each column, decide the most appropriate `dtype` and `fix` it:

In [ ]:
df.info()

`survived`

In [151]:
df['survived'].value_counts()

survived
0    549
1    342
Name: count, dtype: int64

In [152]:
# could be boolean or category - lets convert to boolean
df['survived'] = df['survived'].astype(bool)
df['survived'].dtype

dtype('bool')

In [153]:
df['survived'].value_counts()

survived
False    549
True     342
Name: count, dtype: int64

In [154]:
# if you want as category, make sure to use map() or replace() if you want the category value different than the current value
# df['survived'] = df['survived'].map({0: 'no', 1: 'yes'}).astype('category')

# use df['survived'].cat.categories to verify

`fare`

In [155]:
df.survived.value_counts()

survived
False    549
True     342
Name: count, dtype: int64

In [156]:
# fare is currently as object
df['fare']

0         7.25
1      71.2833
2        7.925
3         53.1
4         8.05
        ...   
886       13.0
887       30.0
888      23.45
889        NaN
890       7.75
Name: fare, Length: 891, dtype: object

In [157]:
# lets identify those values that are making the column as object
# pd.to_numeric(df['fare'], errors='coerce') -> tries to convert every value in df['fare'] to a number (float).
# If a value cannot be converted (like '?', 'abc', '£20'), it becomes NaN.
# .isna() returns True where the value is NaN and False where the value is a valid number.
# .notna() is: True where the original value is not NaN and False where it is NaN

"""
In other words, we are filtering for rows where:

the value fails to convert to a number (becomes NaN),
AND at the same time the original value was not already NaN.
"""

mask = pd.to_numeric(df['fare'], errors='coerce').isna() & df['fare'].notna()
df.loc[mask, 'fare']

7       21.075 USD
8      11.1333 USD
10        16.7 USD
11       26.55 USD
28      7.8792 USD
          ...     
855       9.35 USD
858    19.2583 USD
861       11.5 USD
867    50.4958 USD
871    52.5542 USD
Name: fare, Length: 170, dtype: object

In [158]:
# we are going to use .replace() to remove the ' USD' suffix and convert to float
df['fare'] = (
    df['fare']
        .str.replace(' USD', '')
        .astype(float)
)
df['fare'].dtype

dtype('float64')

In [159]:
# check again
mask = pd.to_numeric(df['fare'], errors='coerce').isna() & df['fare'].notna()
df.loc[mask, 'fare']

Series([], Name: fare, dtype: float64)

`age`

In [160]:
# age as float?
df.age.value_counts()

age
24.00    30
22.00    27
18.00    26
28.00    25
30.00    25
         ..
24.50     1
0.67      1
0.42      1
34.50     1
74.00     1
Name: count, Length: 88, dtype: int64

In [161]:
# if we do not want to see the NaNs
filter_notna = df['age'].notna()
filter_decimals = df['age'] % 1 != 0
df[filter_notna & filter_decimals]['age']

57     28.50
78      0.83
111    14.50
116    70.50
122    32.50
123    32.50
148    36.50
152    55.50
153    40.50
203    45.50
227    20.50
296    23.50
305     0.92
331    45.50
469     0.75
525    40.50
644     0.75
676    24.50
735    28.50
755     0.67
767    30.50
803     0.42
814    30.50
831     0.83
843    34.50
Name: age, dtype: float64

In [162]:
df[df['age'] % 1 != 0]['age'].value_counts()

age
28.50    2
0.83     2
32.50    2
40.50    2
45.50    2
0.75     2
30.50    2
70.50    1
36.50    1
14.50    1
20.50    1
55.50    1
0.92     1
23.50    1
24.50    1
0.67     1
0.42     1
34.50    1
Name: count, dtype: int64

In [163]:
# we will ideally convert age to int... notice that this will reduce the granularity about the babys...
# we also need to decide what we are going to do about the NaNs
df['age'].isna().sum()

np.int64(177)

In [164]:
# lets truncate each value ... you could also round() the value
# there are many ways you can do this...
# in this example, we are going to use our filter previously used to inspect the age column and obtain the rows index where there is a problem
filter_notna = df['age'].notna()
filter_decimals = df['age'] % 1 != 0
df[filter_notna & filter_decimals].index

Index([ 57,  78, 111, 116, 122, 123, 148, 152, 153, 203, 227, 296, 305, 331,
       469, 525, 644, 676, 735, 755, 767, 803, 814, 831, 843],
      dtype='int64')

In [165]:
# now lets fix just those age values
problematic_age = list(df[filter_notna & filter_decimals].index)

# remember that we can use loc to filter
df.loc[57, 'age']

np.float64(28.5)

In [ ]:
df.iloc[list(df['sex']=='Male'), 2]

In [166]:
for age_index in problematic_age:
    df.loc[age_index, 'age'] = int(df.loc[age_index, 'age'])

# this will truncate but the column dtype is still float because there are other floats already

In [167]:
df.loc[57, 'age']


np.float64(28.0)

In [168]:
df['age'] = df['age'].astype('Int64')
df['age'].dtype

Int64Dtype()

In [169]:
df['age']

0        22
1        38
2        26
3        35
4        35
       ... 
886      27
887      19
888    <NA>
889      26
890      32
Name: age, Length: 891, dtype: Int64

**👉 notice that we have missing values in this column... the dtype might be what we want but we can still clean this column**

## Standardizing Categorical Values

Sometimes you may also find inconsistent or incorrect labels in categorical columns.  
For example, the same category may appear in multiple formats:

- "female", "Female", "F", "f"
- "male", "Male", "M", "m"

These inconsistencies make analysis, filtering, and grouping unreliable.  
To fix this, we **standardize** all categorical values into a consistent format.

---
**1. Convert everything to lowercase**
>
>This removes differences like "Female" vs "female".
>
>```python
>df["column_name"] = df["column_name"].astype(str).str.lower()
>```
---
**2. If there are many variations, a `mapping dictionary` might help:**
>
>```python
>mapping = {
>    "female": "female",
>    "f": "female",
>    "woman": "female",
>    "male": "male",
>    "m": "male",
>    "man": "male"
>}
>
>df["column_name"] = df["column_name"].replace(mapping)
>```
___
**Tips:**
>
>- Verify the frequencies of `categorical columns`
>```python
>df["column_name"].value_counts()
>```
>- Converting to `category` offers optimization in terms of memory footprint, comparinson speed and interpretability
>```python
>df["column_name"] = df["column_name"].astype("category")
>```

### Q.2.3 Verify if any categorical column needs to be standardized
- Inspect columns of dtype `category`, `object`, `bool`, `int`
- Inspect using `df['column'].values_count()`

In [324]:
df['sex'].value_counts()

sex
Male      577
female    314
Name: count, dtype: int64

In [325]:
# lets convert all values to lowercase
'Male'.lower()

'male'

In [ ]:
# test the convertion before applying
df['sex'].astype(str).str.lower()

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: object

**now we can apply it with**

```python
df['sex'] = df['sex'].astype(str).str.lower()
```

**but lets suppose that we had more problems...**

**let's use a mapping dictionary**

In [330]:
# we will test before applying it
mapping_dict = {
    'Male': 'male',
    'Female': 'female',
    'Other': 'other',
    'M': 'male',
    'F': 'female',
    'O': 'other',
    np.nan: 'not specified'
}


df['sex'].replace(mapping_dict)

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: object

In [331]:
df['sex'] = df['sex'].replace(mapping_dict)
df['sex'].value_counts()

sex
male      577
female    314
Name: count, dtype: int64

In [ ]:
# the dtype is O which stands for object
df['sex'].dtype

dtype('O')

In [335]:
# it would not hurt to cast this column to category, right?
df['sex'] = df['sex'].astype('category')
df['sex']

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: category
Categories (2, object): ['female', 'male']

## Missing values

Most real-world datasets contain missing or inconsistent values.

We must decide how to handle them.

**The main strategies are:**
___
**1. Remove rows with missing values**

>Useful when only a few rows contain missing values.
>
>- Removing them does not distort the dataset.
>
>```python
>df_clean = df.dropna()
>```
___
**2. Remove columns with too many missing values**

>Useful when a column is mostly empty (e.g., >80% missing maybe even lower threshold)
>
>- The column is not essential for the analysis
>
>```python
>df_clean = df.drop(columns=["column_name"])
>```
___
**3. Impute missing values (numeric columns)**

>Common inputs:
>
>- Mean
>
>- Median (more robust to outliers)
>
>- Constant value
>
>```python
># mean
>df["column_name"] = df["column_name"].fillna(df["column_name"].mean())
>
># median
>df["column_name"] = df["column_name"].fillna(df["column_name"].median())
>
># constant value
>df["column_name"] = df["column_name"].fillna(1)
>```
___
**4. Impute missing values (categorical columns)**

>Typical strategies:
>
>Replace with the most frequent category (mode)
>
>Replace with a special label such as "Unknown"
>
>
>```python
># mode
>df["column_name"] = df["column_name"].fillna(df["column_name"].mode()[0])
>
># special label
>df["column_name"] = df["column_name"].fillna("Unknown")
>```
___
**There are more advanced methods of imputation but those above cover most of the problems that you might have initially.**

### Q.2.4 Check for missing values in every column:
- Try to obtain a sum of missing values in each column

In [218]:
# your code here
df.isnull().sum()

survived            0
pclass              0
sex                 0
age                 0
sibsp               0
parch               0
fare                0
adult_male          0
cabin             688
embark_town         0
passengerId         0
name                0
fare_bin            0
age_fare_ratio     14
surname             0
class               0
dtype: int64

### Q.2.5 Which column is the most problematic in terms of missing values?

In [ ]:
# you answer here
# deck

### Q.2.6 Let's find out the percentage of missing values in the most problematic column.

In [ ]:
# your code here
(df['deck'].isnull().sum() / len(df['deck'])) * 100

np.float64(77.21661054994388)

### Q.2.7 Define a function that will check for missing values in percentage each column that you provide as a param.

**Example:**

```python
def percentage_missing_values(df: pd.DataFrame, col: str) -> float:

    # your code here
    
    return percentage_of_missing
```


In [174]:
# your code here
def percentage_missing_values(df: pd.DataFrame, col: str) -> float:
    return (df[col].isnull().sum() / len(df[col])) * 100

percentage_missing_values(df, 'cabin')

np.float64(77.21661054994388)

### Q 2.8. Impute missing values for each column that requires cleaning

`age`

In [175]:
df['age'].fillna(df['age'].median(), inplace=True)
df['age'].isnull().sum()

np.int64(0)

In [ ]:
# another way to solve this problem would be
df['age'] = (
    pd.to_numeric(df['age'], errors='coerce')           # convert to numeric
      .fillna(df['age'].median())                       # imput NaN with median
      .astype(int)                                      # trunc
)

# one-liner: df['age'] = pd.to_numeric(df['age'], errors='coerce').fillna(df['age'].median()).astype(int)

# much easier :D pay attention to the pipeline but hadling missing values is covered on the following section

`fare`

In [ ]:
# lets input the median too
df['fare'].isnull().sum()

np.int64(45)

In [178]:
df['fare'].fillna(df['fare'].median(), inplace=True)
df['fare'].isnull().sum()

np.int64(0)

**- 1. we could be more specific and input the median per ticket class...**

**- 2. we could be even more specific and also consider the port where the passanger embarked**

**Try any of that!!!** (you migh need to reload the dataset to obtain the missing values again)

`embark_town`

In [181]:
df['embark_town'].isnull().sum()

np.int64(2)

In [ ]:
df['embark_town'].value_counts()

embark_town
Southampton    644
Cherbourg      168
Queenstown      77
Name: count, dtype: int64

In [ ]:
df[df['embark_town'].isnull()]

survived  pclass     sex  age  sibsp  parch  fare  adult_male cabin  \
61       True       1  female   38      0      0  80.0       False     B   
829      True       1  female   62      0      0  80.0       False     B   

    embark_town      passengerId          name  
61          NaN  T1-1912-U-P0001   Kelly Smith  
829         NaN  T1-1912-U-P0002  Penny Miller

In [ ]:
df.loc[df['embark_town'].isnull(), 'embark_town']

61     NaN
829    NaN
Name: embark_town, dtype: object

In [192]:
df.loc[df['embark_town'].isnull(), 'embark_town'] = 'Unknown'

In [194]:
# check again
df[df['embark_town'].isnull()]

Empty DataFrame
Columns: [survived, pclass, sex, age, sibsp, parch, fare, adult_male, cabin, embark_town, passengerId, name]
Index: []

In [196]:
df['embark_town'].isnull().sum()

np.int64(0)

`deck`

In [ ]:
# there is over 77% of missing data... there is nothing we could do right now to save this column :D
# there are advanced techniques that rely on building models that based on other features could infer those values but we won't dwelve deep on this
# and the success of such approach would need to be considered with some discretion

## 3. Feature Engineering

Now that the dataset is clean, with corrected data types, standardized categorical values, and missing values imputed, we can begin improving the dataset by **creating new features**.

Feature Engineering helps models extract more meaningful patterns by transforming the raw data into something more informative and more useful for analysis or machine learning.

---
>**1.Mathematical combinations**
>Create new features by combining existing numeric features.
>
>**Example: family size**
>```python
>df["family_size"] = df["sibsp"] + df["parch"] + 1
>```
---
>**2. Binarizing or grouping categories**
>Transform categories into simpler or grouped labels.
>
>**Example: traveling alone vs not**
>```python
>df["is_alone"] = (df["family_size"] == 1)
>```
---
>**3. Extracting information from existing columns**
>
>**Example: title extracted from the passenger name**
>```python
>df["title"] = df["name"].str.extract(r",\s*([^\.]*)\.", expand=False)
>df["title"] = df["title"].str.strip().str.lower()
>```
---
>**4. Discretization (binning)**
>Convert continuous variables into intervals.
>
>**Example using `cut`:**
>```python
>df["age_group"] = pd.qcut(
>    df["age"],
>    bins=[0, 12, 18, 35, 60, 100],
>    labels=["child", "teen", "young_adult", "adult", "senior"]
>)
>```
>**-  `pd.cut` will consider intervals as `( ]`**
>**- `open on left (exclusive)` and `closed on right (inclusive)`**
>
>**Example using `qcut` (data will be split in equal parts)**
>```python
>df["age_group"] = pd.cut(
>    df["age"],
>    q=5,
>    labels=["child", "teen", "young_adult", "adult", "senior"]
>)
>```
---
>**5. Encoding boolean or numeric signals**
>
>**Example: indicator for high fare**
>```python
>df["high_fare"] = df["fare"] > df["fare"].median()
>```
---
>**6. Interaction features**
>Features created by combining multiple existing ones.
>
>**Example: price paid per person**
>```python
>df["fare_per_person"] = df["fare"] / df["family_size"]
>```
---
**Example Creating feature set**
>
>```python
>df["family_size"] = df["sibsp"] + df["parch"] + 1
>df["is_alone"] = (df["family_size"] == 1)
>
>df["title"] = df["name"].str.extract(r",\s*([^\.]*)\.", expand=False)
>df["title"] = df["title"].str.strip().str.lower()
>
>df["age_group"] = pd.cut(
>    df["age"],
>    bins=[0, 12, 18, 35, 60, 100],
>    labels=["child", "teen", "young_adult", "adult", "senior"]
>)
>
>df["high_fare"] = df["fare"] > df["fare"].median()
>df["fare_per_person"] = df["fare"] / df["family_size"]
>
>df.head()
>```
---
### Q 3.1. Create a new column called `is_child` that is `True` when **age < 12**.

In [220]:
df['is_child'] = df['age'] < 12
df['is_child']

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: is_child, Length: 891, dtype: boolean

### **Q 3.2.** Create a column `family_size` based on sibsp e parch.

In [222]:
df['family_size'] = df['sibsp'] + df['parch'] + 1
df['family_size']

0      2
1      2
2      1
3      2
4      1
      ..
886    1
887    1
888    4
889    1
890    1
Name: family_size, Length: 891, dtype: int64

### Q 3.3. Create a `class` column based on the `Ticket Class`.
- **1 -> First**
- **2 -> Second**
- **3 -> Third**

- **Set the `dtype` of this column to `category`.**

In [206]:
df['class'] = df['pclass'].replace({
    1: 'First',
    2: 'Second',
    3: 'Third'
}).astype('category')

### Q.3.4. Create a `embarked` column based on the Port where the passenger embarked.
- **Use the initials as the category value.**
- **Set the `dtype` of this column to `category`.**

In [229]:
df['embarked'] = df['embark_town'].str[0]
df['embarked']

0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: embarked, Length: 891, dtype: object

In [230]:
df['embarked'].value_counts()

embarked
S    644
C    168
Q     77
U      2
Name: count, dtype: int64

### Q 3.5. Create a `fare_bin` column by splitting fare into 4 quantiles using `pd.qcut`.
- use `labels` as a list of `'Low', 'Medium', 'High', 'Very High'`

In [202]:
df['fare_bin'] = pd.qcut(df['fare'], q=4, labels=['Low', 'Medium', 'High', 'Very High'])
df['fare_bin']

0            Low
1      Very High
2            Low
3      Very High
4         Medium
         ...    
886       Medium
887    Very High
888         High
889       Medium
890          Low
Name: fare_bin, Length: 891, dtype: category
Categories (4, object): ['Low' < 'Medium' < 'High' < 'Very High']

### Q 3.6. Create a column `age_fare_ratio` equal to `age / fare`, and comment on whether this feature makes sense.
- be careful about zero-division

In [ ]:
# the replace is to avoid zero-division
df['age_fare_ratio'] = df['age'] / df['fare'].replace(0, np.nan)
df['age_fare_ratio']

0      3.034483
1      0.533084
2      3.280757
3      0.659134
4      4.347826
         ...   
886    2.076923
887    0.633333
888     1.19403
889    1.798785
890    4.129032
Name: age_fare_ratio, Length: 891, dtype: Float64

In [ ]:
# lets verify if there is null values...
df['age_fare_ratio'].isnull().sum()

np.int64(14)

**👉 Important**
- **we just derived this feature - creating it introduced some nans that we could clean. Perhaps based on some strategie we used before or a new one...**
- **check the next question... maybe we could use it and input the median of `age_fare_ratio` based on the `age_group`**
- **try it after the `Q.3.7`**

### Q 3.7. Create `age_group` **(you decide the bins and labels)** and count how many passengers exist in each `age_group`.

In [292]:
df['age_group'] = pd.cut(df['age'], bins=[0, 11, 18, 30, 50, 70], labels=['child', 'teenager', 'young adult', 'adult', 'senior'])
df['age_group'].value_counts()

age_group
young adult    449
adult          239
teenager        71
child           61
senior          60
Name: count, dtype: int64

In [ ]:
# it seems correct right? but what if there were passengers with age above 70?
# what about the passengers with 0 years?
# notice the NaN when we verify for the unique values. .values_count() does not show NaNs...

df['age_group'].unique()

['young adult', 'adult', 'senior', 'child', 'teenager', NaN]
Categories (5, object): ['child' < 'teenager' < 'young adult' < 'adult' < 'senior']

In [266]:
df[df['age'] > 70][['age', 'age_group']]

age age_group
96    71       NaN
493   71       NaN
630   80       NaN
851   74       NaN

In [267]:
df[df['age'] == 0][['age', 'age_group']]

age age_group
78     0       NaN
305    0       NaN
469    0       NaN
644    0       NaN
755    0       NaN
803    0       NaN
831    0       NaN

In [280]:
max(df['age'])

np.int64(80)

**let's make sure that we catch all the intervals now**

**we can use the -1 on the left since there is no age below 0 and the max(df['age'])**

**we could also discretize the adult and senior interval a bit more**

In [293]:
df['age_group'] = pd.cut(df['age'], bins=[-1, 3, 11, 18, 30, 50, 65, max(df['age'])], labels=['baby', 'child', 'teenager', 'young adult', 'adult', 'mature', 'senior'],)
df['age_group'].value_counts()

age_group
young adult    449
adult          239
teenager        71
mature          56
child           38
baby            30
senior           8
Name: count, dtype: int64

**alternatively, we can use the `float(-inf)` and `float('inf')`**

In [ ]:
df['age_group'] = pd.cut(df['age'], bins=[float('-inf'), 3, 11, 18, 30, 50, 65, float('inf')], labels=['baby', 'child', 'teenager', 'young adult', 'adult', 'mature', 'senior'])
df['age_group'].value_counts()

age_group
young adult    449
adult          239
teenager        71
mature          56
child           38
baby            30
senior           8
Name: count, dtype: int64

In [ ]:
# now all babys are covered
len(df[(df['age'] >= 0) & (df['age'] <= 3)])

30

In [ ]:
# 30 babys + 38 childs == 68 is_child
df['is_child'].value_counts()

is_child
False    823
True      68
Name: count, dtype: Int64

In [ ]:
# and all seniors too
df[df['age'] > 70][['age', 'age_group']]

age age_group
96    71    senior
493   71    senior
630   80    senior
851   74    senior

### Q 3.8. Create a new column `surname` containing the passenger’s surname

In [248]:
df['surname'] = df['name'].str.split().str[-1]
df['surname']

0        Flores
1         Tyler
2       Mendoza
3       Pearson
4        Taylor
         ...   
886     Johnson
887        Buck
888    Lawrence
889        Reed
890      Horton
Name: surname, Length: 891, dtype: object

### **Q 3.9.** Using the engineered features, compute a summary showing:
- average `fare` per `age_group`
- average `family_size` per `class`

average `fare` per `age_group`

In [297]:
df['age_group'].value_counts()

age_group
young adult    449
adult          239
teenager        71
mature          56
child           38
baby            30
senior           8
Name: count, dtype: int64

In [ ]:
# lets check again the categories as a list
list(df['age_group'].cat.categories)

['baby', 'child', 'teenager', 'young adult', 'adult', 'mature', 'senior']

In [303]:
# we could also use unique but it would not respect the category order...
list(df['age_group'].unique())

['young adult', 'adult', 'mature', 'baby', 'teenager', 'child', 'senior']

In [ ]:
# first try to build the filter - then you can generalize
df[df['age_group'] == 'baby']['fare'].mean()

np.float64(33.00153)

In [ ]:
# now we will use a variable instead of the baby category and iterate on it
for group in list(df['age_group'].cat.categories):
    print(f"Average Fare of {group}: {df.loc[df['age_group'] == group, 'fare'].mean()}")

Average Fare of baby: 33.00153
Average Fare of child: 29.021163157894733
Average Fare of teenager: 29.728584507042253
Average Fare of young adult: 24.385290868596883
Average Fare of adult: 41.14185983263599
Average Fare of mature: 41.23095357142857
Average Fare of senior: 27.710425


In [307]:
# without loc
for group in list(df['age_group'].cat.categories):
    print(f"Average Fare of {group}: {df[df['age_group'] == group]['fare'].mean()}")

Average Fare of baby: 33.00153
Average Fare of child: 29.021163157894733
Average Fare of teenager: 29.728584507042253
Average Fare of young adult: 24.385290868596883
Average Fare of adult: 41.14185983263599
Average Fare of mature: 41.23095357142857
Average Fare of senior: 27.710425


average `family_size` per `class`

In [309]:
df['class'].cat.categories

Index(['First', 'Second', 'Third'], dtype='object')

In [ ]:
# build your filter for one category then generalize
df.loc[df['class'] == 'First', 'family_size'].mean()

np.float64(1.7731481481481481)

In [316]:
for group in list(df['class'].cat.categories):
    print(f"Average Family Size of {group}: {df.loc[df['class'] == group, 'family_size'].mean()}")

Average Family Size of First: 1.7731481481481481
Average Family Size of Second: 1.7826086956521738
Average Family Size of Third: 2.0081466395112018


In [315]:
# without loc
for group in list(df['class'].cat.categories):
    print(f"Average Family Size of {group}: {df[df['class'] == group]['family_size'].mean()}")

Average Family Size of First: 1.7731481481481481
Average Family Size of Second: 1.7826086956521738
Average Family Size of Third: 2.0081466395112018


## 4. Export Dataset

Now that our dataset is clean, we might want to save time by not needing to preprocess it again.

Exporting it is a good way to avoid running all cleaning steps every time you start your notebook.

There are many formats you can choose, and the most common are:

---

### ✔️ Export to CSV
>```python
>df_clean.to_csv("titanic_clean.csv", index=False)
>```
>
>- `index=False` avoids saving the DataFrame index as an extra column.
>- CSV files are human-readable and compatible with most tools.
---
### ✔️ Export to Excel
>```python
>df_clean.to_excel("titanic_clean.xlsx", index=False)
>```
>
>- Requires `openpyxl` or `xlsxwriter` installed.
>- Useful when sharing data with non-programmers.
---
### ✔️ Export to JSON
>```python
>df_clean.to_json("titanic_clean.json", orient="records")
>```
>
>**Common orientations:**
>
>- `"records"` → list of dictionaries (most common)
>- `"split"` → dictionary with `index`, `columns`, `data`
>- `"table"` → includes metadata, useful for interoperability
---
### ✔️ Loading Back the Dataset Later
>```python
>df_loaded = pd.read_csv("titanic_clean.csv")
>```
>
>**or:**
>
>```python
>df_loaded = pd.read_json("titanic_clean.json")
>```
---
**if you are curious, search for `pickle` and `parquet`.**

**Now you can skip all preprocessing and jump directly into analysis.**

### Q 4.1. Export the dataset to `csv`. Set `index=True` as an additional param.

In [317]:
df.to_csv('cleaned_titanic.csv', index=True)

### Q 4.2 Load into a variable named `df_loaded`

In [318]:
# be careful, do not overwrite your current dataframe variable
df_loaded = pd.read_csv('cleaned_titanic.csv')
df_loaded

Unnamed: 0  survived  pclass     sex  age  sibsp  parch     fare  \
0             0     False       3    Male   22      1      0   7.2500   
1             1      True       1  female   38      1      0  71.2833   
2             2      True       3  female   26      0      0   7.9250   
3             3      True       1  female   35      1      0  53.1000   
4             4     False       3    Male   35      0      0   8.0500   
..          ...       ...     ...     ...  ...    ...    ...      ...   
886         886     False       2    Male   27      0      0  13.0000   
887         887      True       1  female   19      0      0  30.0000   
888         888     False       3  female   28      1      2  23.4500   
889         889      True       1    Male   26      0      0  14.4542   
890         890     False       3    Male   32      0      0   7.7500   

     adult_male cabin  ...      passengerId                 name   fare_bin  \
0          True   NaN  ...  T3-1912-S-P0001        Dustin Flores        Low   
1         False     C  ...  T1-1912-C-P0001           Mary Tyler  Very High   
2         False   NaN  ...  T3-1912-S-P0002         Anna Mendoza        Low   
3         False     C  ...  T1-1912-S-P0001        Debra Pearson  Very High   
4          True   NaN  ...  T3-1912-S-P0003       Anthony Taylor     Medium   
..          ...   ...  ...              ...                  ...        ...   
886        True   NaN  ...  T2-1912-S-P0164      Michael Johnson     Medium   
887       False     B  ...  T1-1912-S-P0127            Lisa Buck  Very High   
888       False   NaN  ...  T3-1912-S-P0353  Alexandria Lawrence       High   
889        True     C  ...  T1-1912-C-P0085     Christopher Reed     Medium   
890        True   NaN  ...  T3-1912-Q-P0072         James Horton        Low   

    age_fare_ratio   surname   class is_child  family_size  embarked  \
0         3.034483    Flores   Third    False            2         S   
1         0.533084     Tyler   First    False            2         C   
2         3.280757   Mendoza   Third    False            1         S   
3         0.659134   Pearson   First    False            2         S   
4         4.347826    Taylor   Third    False            1         S   
..             ...       ...     ...      ...          ...       ...   
886       2.076923   Johnson  Second    False            1         S   
887       0.633333      Buck   First    False            1         S   
888       1.194030  Lawrence   Third    False            4         S   
889       1.798785      Reed   First    False            1         C   
890       4.129032    Horton   Third    False            1         Q   

       age_group  
0    young adult  
1          adult  
2    young adult  
3          adult  
4          adult  
..           ...  
886  young adult  
887  young adult  
888  young adult  
889  young adult  
890        adult  

[891 rows x 21 columns]

In [ ]:
# that's why we usually dont save with the index set to True

**You also could just drop the artificial index column...** `df.drop(columns=['col'], inplace=True)`

### Q 4.3 Export the dataset to `csv`. Set `index=False` as an additional param.

In [336]:
df.to_csv('cleaned_titanic.csv', index=False)

## 5. Final Clean Dataset

### Load the Clean Dataset

- **Test first with a different variable name to avoid overwriting by accident.**

In [337]:
df_loaded = pd.read_csv('cleaned_titanic.csv')
df_loaded

survived  pclass     sex  age  sibsp  parch     fare  adult_male deck  \
0       False       3    male   22      1      0   7.2500        True  NaN   
1        True       1  female   38      1      0  71.2833       False    C   
2        True       3  female   26      0      0   7.9250       False  NaN   
3        True       1  female   35      1      0  53.1000       False    C   
4       False       3    male   35      0      0   8.0500        True  NaN   
..        ...     ...     ...  ...    ...    ...      ...         ...  ...   
886     False       2    male   27      0      0  13.0000        True  NaN   
887      True       1  female   19      0      0  30.0000       False    B   
888     False       3  female   28      1      2  23.4500       False  NaN   
889      True       1    male   26      0      0  14.4542        True    C   
890     False       3    male   32      0      0   7.7500        True  NaN   

     embark_town      passengerId                 name   fare_bin  \
0    Southampton  T3-1912-S-P0001        Dustin Flores        Low   
1      Cherbourg  T1-1912-C-P0001           Mary Tyler  Very High   
2    Southampton  T3-1912-S-P0002         Anna Mendoza        Low   
3    Southampton  T1-1912-S-P0001        Debra Pearson  Very High   
4    Southampton  T3-1912-S-P0003       Anthony Taylor     Medium   
..           ...              ...                  ...        ...   
886  Southampton  T2-1912-S-P0164      Michael Johnson     Medium   
887  Southampton  T1-1912-S-P0127            Lisa Buck  Very High   
888  Southampton  T3-1912-S-P0353  Alexandria Lawrence       High   
889    Cherbourg  T1-1912-C-P0085     Christopher Reed     Medium   
890   Queenstown  T3-1912-Q-P0072         James Horton        Low   

     age_fare_ratio   surname   class  is_child  family_size embarked  \
0          3.034483    Flores   Third     False            2        S   
1          0.533084     Tyler   First     False            2        C   
2          3.280757   Mendoza   Third     False            1        S   
3          0.659134   Pearson   First     False            2        S   
4          4.347826    Taylor   Third     False            1        S   
..              ...       ...     ...       ...          ...      ...   
886        2.076923   Johnson  Second     False            1        S   
887        0.633333      Buck   First     False            1        S   
888        1.194030  Lawrence   Third     False            4        S   
889        1.798785      Reed   First     False            1        C   
890        4.129032    Horton   Third     False            1        Q   

       age_group  
0    young adult  
1          adult  
2    young adult  
3          adult  
4          adult  
..           ...  
886  young adult  
887  young adult  
888  young adult  
889  young adult  
890        adult  

[891 rows x 20 columns]